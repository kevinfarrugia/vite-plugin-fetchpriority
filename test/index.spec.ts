import { resolve } from "node:path";

import { OutputAsset, RollupOutput } from "rollup";
import { build } from "vite";
import { test, describe, expect } from "vitest";
import fetchPriority, { FetchPriorityOptions } from "../src";

async function buildVite(options: FetchPriorityOptions[]) {
  const { output } = (await build({
    root: resolve(__dirname, "./fixtures"),
    plugins: [fetchPriority(options)],
    build: {
      rollupOptions: {
        output: {
          manualChunks: (id) => {
            // separate chunk for each file
            if (id.endsWith(".js")) {
              return id.split("/").pop()?.replace(".js", "");
            }
            return null;
          },
        },
      },
    },
  })) as RollupOutput;

  const { source: indexSource } = output.find(
    (item) => item.fileName === "index.html"
  ) as OutputAsset;

  return indexSource.toString();
}

describe("fetchPriority", () => {
  test("should not add fetchpriority attribute if no matching files are found", async () => {
    const options: FetchPriorityOptions[] = [
      {
        priority: "high",
        files: [
          {
            type: "script",
            match: /nonexistent/,
          },
        ],
      },
    ];

    await buildVite(options).then((result) => {
      expect(result).toMatchSnapshot();
    });
  });

  test("should add fetchpriority=high to all script tags", async () => {
    const options: FetchPriorityOptions[] = [
      {
        priority: "high",
        files: [
          {
            type: "script",
          },
        ],
      },
    ];

    await buildVite(options).then((result) => {
      expect(result).toMatchSnapshot();
    });
  });

  test("should add fetchpriority=high to style tags", async () => {
    const options: FetchPriorityOptions[] = [
      {
        priority: "high",
        files: [
          {
            type: "style",
          },
        ],
      },
    ];

    await buildVite(options).then((result) => {
      expect(result).toMatchSnapshot();
    });
  });

  test("should add fetchpriority=high to style tags that match RegEx", async () => {
    const options: FetchPriorityOptions[] = [
      {
        priority: "high",
        files: [
          {
            type: "style",
            match: /test/,
          },
        ],
      },
    ];

    await buildVite(options).then((result) => {
      expect(result).toMatchSnapshot();
    });
  });

  test("should add fetchpriority=low to style tags that match RegEx", async () => {
    const options: FetchPriorityOptions[] = [
      {
        priority: "high",
        files: [
          {
            type: "script",
            match: /app/,
          },
        ],
      },
      {
        priority: "low",
        files: [
          {
            type: "style",
            match: /test/,
          },
        ],
      },
    ];

    await buildVite(options).then((result) => {
      expect(result).toMatchSnapshot();
    });
  });

  test("should add fetchpriority=high to link rel=preload", async () => {
    const options: FetchPriorityOptions[] = [
      {
        priority: "high",
        files: [
          {
            type: "preload",
          },
        ],
      },
    ];

    await buildVite(options).then((result) => {
      expect(result).toMatchSnapshot();
    });
  });

  test("should handle multiple priorities for the same file type", async () => {
    const options: FetchPriorityOptions[] = [
      {
        priority: "high",
        files: [
          {
            type: "script",
          },
        ],
      },
      {
        priority: "low",
        files: [
          {
            type: "script",
            match: /random/,
          },
        ],
      },
    ];

    await buildVite(options).then((result) => {
      expect(result).toMatchSnapshot();
    });
  });

  test("should add fetchpriority=high to multiple file types", async () => {
    const options: FetchPriorityOptions[] = [
      {
        priority: "high",
        files: [
          {
            type: "script",
          },
          {
            type: "style",
          },
          {
            type: "preload",
          },
        ],
      },
    ];

    await buildVite(options).then((result) => {
      expect(result).toMatchSnapshot();
    });
  });
});
