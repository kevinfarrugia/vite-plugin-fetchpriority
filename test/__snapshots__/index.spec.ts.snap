// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fetchPriority > should add fetchpriority=high to all script tags 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
    <link rel="stylesheet" href="/app.css" />
    <link rel="stylesheet" href="/test.css" />
    <link rel="preload" href="/roboto.woff2" type="font/woff2" as="font" crossorigin="anonymous" />
    <link rel="preload" href="/roboto-semibold.woff2" type="font/woff2" as="font" crossorigin="anonymous" />
    <script type="module" crossorigin src="/assets/modulepreload-polyfill-B5Qt9EMX.js" fetchpriority="high"></script>
    <script type="module" crossorigin src="/assets/app-l0sNRNKZ.js" fetchpriority="high"></script>
    <script type="module" crossorigin src="/assets/random-DnnRr8WQ.js" fetchpriority="high"></script>
  </head>
  <body>
    <h1>Test App</h1>
  </body>
</html>"
`;

exports[`fetchPriority > should add fetchpriority=high to link rel=preload 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
    <link rel="stylesheet" href="/app.css" />
    <link rel="stylesheet" href="/test.css" />
    <link rel="preload" href="/roboto.woff2" type="font/woff2" as="font" crossorigin="anonymous" / fetchpriority="high">
    <link rel="preload" href="/roboto-semibold.woff2" type="font/woff2" as="font" crossorigin="anonymous" / fetchpriority="high">
    <script type="module" crossorigin src="/assets/modulepreload-polyfill-B5Qt9EMX.js"></script>
    <script type="module" crossorigin src="/assets/app-l0sNRNKZ.js"></script>
    <script type="module" crossorigin src="/assets/random-DnnRr8WQ.js"></script>
  </head>
  <body>
    <h1>Test App</h1>
  </body>
</html>"
`;

exports[`fetchPriority > should add fetchpriority=high to multiple file types 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
    <link rel="stylesheet" href="/app.css" / fetchpriority="high">
    <link rel="stylesheet" href="/test.css" / fetchpriority="high">
    <link rel="preload" href="/roboto.woff2" type="font/woff2" as="font" crossorigin="anonymous" / fetchpriority="high">
    <link rel="preload" href="/roboto-semibold.woff2" type="font/woff2" as="font" crossorigin="anonymous" / fetchpriority="high">
    <script type="module" crossorigin src="/assets/modulepreload-polyfill-B5Qt9EMX.js" fetchpriority="high"></script>
    <script type="module" crossorigin src="/assets/app-l0sNRNKZ.js" fetchpriority="high"></script>
    <script type="module" crossorigin src="/assets/random-DnnRr8WQ.js" fetchpriority="high"></script>
  </head>
  <body>
    <h1>Test App</h1>
  </body>
</html>"
`;

exports[`fetchPriority > should add fetchpriority=high to style tags 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
    <link rel="stylesheet" href="/app.css" / fetchpriority="high">
    <link rel="stylesheet" href="/test.css" / fetchpriority="high">
    <link rel="preload" href="/roboto.woff2" type="font/woff2" as="font" crossorigin="anonymous" />
    <link rel="preload" href="/roboto-semibold.woff2" type="font/woff2" as="font" crossorigin="anonymous" />
    <script type="module" crossorigin src="/assets/modulepreload-polyfill-B5Qt9EMX.js"></script>
    <script type="module" crossorigin src="/assets/app-l0sNRNKZ.js"></script>
    <script type="module" crossorigin src="/assets/random-DnnRr8WQ.js"></script>
  </head>
  <body>
    <h1>Test App</h1>
  </body>
</html>"
`;

exports[`fetchPriority > should add fetchpriority=high to style tags that match RegEx 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
    <link rel="stylesheet" href="/app.css" />
    <link rel="stylesheet" href="/test.css" / fetchpriority="high">
    <link rel="preload" href="/roboto.woff2" type="font/woff2" as="font" crossorigin="anonymous" />
    <link rel="preload" href="/roboto-semibold.woff2" type="font/woff2" as="font" crossorigin="anonymous" />
    <script type="module" crossorigin src="/assets/modulepreload-polyfill-B5Qt9EMX.js"></script>
    <script type="module" crossorigin src="/assets/app-l0sNRNKZ.js"></script>
    <script type="module" crossorigin src="/assets/random-DnnRr8WQ.js"></script>
  </head>
  <body>
    <h1>Test App</h1>
  </body>
</html>"
`;

exports[`fetchPriority > should add fetchpriority=low to style tags that match RegEx 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
    <link rel="stylesheet" href="/app.css" />
    <link rel="stylesheet" href="/test.css" / fetchpriority="low">
    <link rel="preload" href="/roboto.woff2" type="font/woff2" as="font" crossorigin="anonymous" />
    <link rel="preload" href="/roboto-semibold.woff2" type="font/woff2" as="font" crossorigin="anonymous" />
    <script type="module" crossorigin src="/assets/modulepreload-polyfill-B5Qt9EMX.js"></script>
    <script type="module" crossorigin src="/assets/app-l0sNRNKZ.js" fetchpriority="high"></script>
    <script type="module" crossorigin src="/assets/random-DnnRr8WQ.js"></script>
  </head>
  <body>
    <h1>Test App</h1>
  </body>
</html>"
`;

exports[`fetchPriority > should handle multiple priorities for the same file type 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
    <link rel="stylesheet" href="/app.css" />
    <link rel="stylesheet" href="/test.css" />
    <link rel="preload" href="/roboto.woff2" type="font/woff2" as="font" crossorigin="anonymous" />
    <link rel="preload" href="/roboto-semibold.woff2" type="font/woff2" as="font" crossorigin="anonymous" />
    <script type="module" crossorigin src="/assets/modulepreload-polyfill-B5Qt9EMX.js" fetchpriority="high"></script>
    <script type="module" crossorigin src="/assets/app-l0sNRNKZ.js" fetchpriority="high"></script>
    <script type="module" crossorigin src="/assets/random-DnnRr8WQ.js" fetchpriority="low"></script>
  </head>
  <body>
    <h1>Test App</h1>
  </body>
</html>"
`;

exports[`fetchPriority > should not add fetchpriority attribute if no matching files are found 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
    <link rel="stylesheet" href="/app.css" />
    <link rel="stylesheet" href="/test.css" />
    <link rel="preload" href="/roboto.woff2" type="font/woff2" as="font" crossorigin="anonymous" />
    <link rel="preload" href="/roboto-semibold.woff2" type="font/woff2" as="font" crossorigin="anonymous" />
    <script type="module" crossorigin src="/assets/modulepreload-polyfill-B5Qt9EMX.js"></script>
    <script type="module" crossorigin src="/assets/app-l0sNRNKZ.js"></script>
    <script type="module" crossorigin src="/assets/random-DnnRr8WQ.js"></script>
  </head>
  <body>
    <h1>Test App</h1>
  </body>
</html>"
`;
